beamer/middleware.py

```diff
@@ -226,50 +226,49 @@ def _rate_limiter_inner(
-                    # We have reached the end of tapering period.
-                    state.taper_counter = 0
-                    state.taper_counter_max = 0
-                    log.debug(
-                        "Exiting tapering mode", thread=threading.current_thread().name, rpc=rpc
-                    )
-
-                time.sleep(request_delay)
-
-            # If we get rate limited by the RPC, start a new rate limiting
-            # period by setting state.rate_limit_end and try sending the
-            # request again after _RATE_LIMIT_REQUEST_DELAY. Otherwise,
-            # simply return the response.
-            rate_limited, response = _try_make_request(make_request, method, params)
-            if rate_limited:
+
+            assert response is not None
+
+            if time.time() > state.rate_limit_end:
+                # Rate limit period ended so we now enter the tapering period,
+                num_waiting_on_lock = state.num_waiting_on_lock.load()
                 log.debug(
-                    "Entering rate limiting mode", thread=threading.current_thread().name, rpc=rpc
+                    "Exiting rate limiting mode",
+                    thread=threading.current_thread().name,
+                    rpc=rpc,
                 )
-                state.rate_limit_end = time.time() + _RATE_LIMIT_PERIOD
-                time.sleep(_RATE_LIMIT_REQUEST_DELAY)
-            else:
-                assert response is not None
-                return response
+                state.rate_limit_end = None
+                state.taper_counter = 0
+                state.taper_counter_max = num_waiting_on_lock
+
+            return response
+
+        if state.taper_counter_max > 0:
+            # We are in tapering period.
+            request_delay = _RATE_LIMIT_REQUEST_DELAY / 2**state.taper_counter
+            state.taper_counter += 1
+            if state.taper_counter == state.taper_counter_max:
+                # We have reached the end of tapering period.
+                state.taper_counter = 0
+                state.taper_counter_max = 0
+                log.debug("Exiting tapering mode", thread=threading.current_thread().name, rpc=rpc)
+
+            time.sleep(request_delay)
+
+        # If we get rate limited by the RPC, start a new rate limiting
+        # period by setting state.rate_limit_end and try sending the
+        # request again after _RATE_LIMIT_REQUEST_DELAY. Otherwise,
+        # simply return the response.
+        rate_limited, response = _try_make_request(make_request, method, params)
+        if rate_limited:
+            log.debug(
+                "Entering rate limiting mode", thread=threading.current_thread().name, rpc=rpc
+            )
+            state.rate_limit_end = time.time() + _RATE_LIMIT_PERIOD
+            time.sleep(_RATE_LIMIT_REQUEST_DELAY)
+        else:
+            assert response is not None
+            return response
```

Comment:
I have no idea what happened here and github at least is not helpful when showing the diff.
This is some of the trickiest code in our agent, could you please explain what you did here?

Reply:
No need to do `else` after `return`